### PR TITLE
Do not call vararg C functions from Fortran (JIRA CGNS-179) #182

### PR DESCRIPTION
**Pull request by MicK7**
>
>Please see rationals on JIRA CGNS-179:
>https://cgnsorg.atlassian.net/browse/CGNS-179
>

Duplicate of #182 but with the branch on main repository

### CODE REVIEW

**Comment by MicK7** [comment]
@pini-gh Your PR is here.

**Comment by brtnfld** [comment]
Extension: cgns_f.F90, line 1492: Line longer than 132 characters
Extension: cgns_f.F90, line 1644: Line longer than 132 characters
Extension: cgns_f.F90, line 1704: Line longer than 132 characters
Extension: cgns_f.F90, line 1735: Line longer than 132 characters
Extension: cgns_f.F90, line 2078: Line longer than 132 characters
Extension: cgns_f.F90, line 2247: Line longer than 132 characters
Extension: cgns_f.F90, line 4926: Line longer than 132 characters
Warning: cgns_f.F90, line 5219: Unused dummy variable NAME20
Warning: cgns_f.F90, line 5381: Unused dummy variable NAME20
Warning: cgns_f.F90, line 5388: Unused dummy variable A
Warning: cgns_f.F90, line 5395: Unused dummy variable A
Warning: cgns_f.F90, line 5402: Unused dummy variable A
Warning: cgns_f.F90, line 5409: Unused dummy variable A

ng: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 300: Inconsistent data type DOUBLE PRECISION (previously REAL) for argument [ 65%] Li
nking C executable write_grid_unst
2 in reference to CG_ARRAY_READ_F
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 367: In[ 66%] Linking C executable write_grid2zn_str
consistent data type DOUBLE PRECISION (previously REAL) for argument 2 in reference to CG_ARRAY_READ_F
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 391: Inconsistent data type DOUBLE PRECISION (previously REAL) for argument 1 in
 reference to CG_EXPONENTS_READ_F
[ 66%] Linking C executable test_bbox
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 409: Inconsistent data type DOUBLE PRECISION (previously REAL) for argument 1 in
 reference to CG_CONVERSION_READ_F
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 472: Inconsistent data type DOUBLE PRECISION (previously REAL) for argument [ 67
%] Linking C executable read_descriptor
2 in reference to CG_ARRAY_READ_F
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 523: Inconsistent data type DOUB[ 69%] Linking C executable write_bcdata
LE PRECISION (pre[ 68%] Linking C executable write_timevert_str
viously REAL) for argum[ 69%] Linking C executable write_flowvert_str
ent [ 69%] Linking C executable read_flowcentrind_str
8 in reference to CG_COORD_READ_F
[ 69%] Linking C executable write_dimensional
[ 70%] Linking C executable cgioc
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 591: Inconsistent data type DOUBLE PRECISION (previously REAL) for argument 2 in
 reference to CG_ARRAY_READ_F
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 707: Inconsistent data type DOUBLE PRECISION (previously REAL) for argument 2 in
 reference to CG_ARRAY_READ_F
[ 71%] Linking C executable write_flowcentrind_str
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 938: Inconsistent data type DOUBLE PRECISION (previously REAL) for argument 2 in
 reference to CG_ARRAY_READ_F
Warning: /home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/cgread_f03.F90, line 986: Inco[ 72%] Linking C executable read_grid2zn_str
nsistent data type DOUBLE PRECISION (previously REAL) for argument 6 in reference to CG_BOCO_READ_F
[ 98%] Linking C executable read_timevert_str
CMakeFiles/test_family_treef.dir/test_family_treef.F90.o: In function `testfamilytreef_':
/home/hdftest/snapshots-cgns/TestDir/jelly/test.1/CGNS_SRC/src/tests/test_family_treef.F90:35: undefined reference to `unlink_'

**Comment by MicK7** [comment]
The line 35 of test_family_treef.F90 can be removed

File: src/cgns_f.F90 [hunk 1]
```diff
@@ -5046,6 +5055,331 @@ END SUBROUTINE cgio_set_label_f
...
+     INTEGER, OPTIONAL :: index17
+     CHARACTER(*), OPTIONAL :: name18
+     INTEGER, OPTIONAL :: index18
+     CHARACTER(*), OPTIONAL :: name19
+     INTEGER, OPTIONAL :: index19
+     CHARACTER(*), OPTIONAL :: name20
+
+     IF (.NOT. PRESENT(index1)) THEN
+        CALL cg_gorel_f1(fn, ier, name1, 0)
+        RETURN
```
Comment (brtnfld) [inline src/cgns_f.F90]: The issue with the RETURNs is you are going to miss the corner case if a user specifies, for example, name2="x" and index=2, which might be unusual but is allowed by the standard.

**Comment by MicK7** [comment]
The maximum depth for a node path has always been defined as  CG_MAX_GOTO_DEPTH = 20 in the cgnslib. Thus, no one should be using node paths greater than that.
I agree with you on the original implementers intent. May be this function does not make sense in Fortran and we should recommand usage of cg_gopath.

**Comment by brtnfld** [comment]
Thanks, I missed the CG_MAX_GOTO_DEPTH = 20. This limit is not reflected in the documentation for the API.

Are you saying we should deprecate cg_goto_f? Couldn't you say the same about cg_goto in general?

**Comment by MicK7** [comment]
I don't think removing it for C is a good thing since vararg are native in C.
For FORTRAN, you can do a lot of things with only cg_gopath_f, cg_golist_f and cg_where_f but in the end cg_goto_f is easier to do quick code writing (even if the API is a bit confusing (mixing node label and node name).
I am just stating that we could promote the usage of cg_gopath_f in the documentation for people using Fortran.
But in this case, maybe an equivalent function to cg_where_f would be needed (like a 'cg_getpath_f' to get current path and depth position)

File: src/cgns_f.F90 (hunk 1 above)
Comment (MicK7) [reply src/cgns_f.F90]: How is this corner case possible ? 
Indeed if index1 is not present, it should imply that name1=="end" and the goto do not change position. The proposed code is not checking that name1 is indeed equal to "end" and is relying on the fact that underlying C code will reach name1  before the other "end" in cg_gorel_f1 (weird choice of coding??) to stop treating the input parameters.

**Comment by MicK7** [comment]
@brtnfld You were right. The current proposed patch is missing the 20th path/index of cg_goto and is not coherent with cg_goto_f08 already existing in cgns_f.F90 file. Moreover, I fear that it breaks CRAY and WINDOWS string transfer from fortran to C. Should we drop this MR ?

**Comment by MicK7** [comment]
@pini-gh This PR should also help with JIRA CGNS-205 issue.
To compile with gfortran-10, the following flags need to be provided "-std=f2008ts -fallow-argument-mismatch"

File: src/tests/cgsubreg_f03.F90 [hunk 2]
```diff
@@ -12,8 +12,8 @@ program cgsubreg
       integer(cgsize_t) ptrange(6), ptlist(125), npts
       integer(cgenum_t) loc, ptype
       integer bclen, gclen
-      real*4 x(125), y(125), z(125)
-      character*32 sname, cname
+      real(kind=4) x(125), y(125), z(125)
```
[the remark this answers is not in the document]
Comment (brtnfld) [reply src/tests/cgsubreg_f03.F90:15]: fixing now.

**Comment by MicK7** [comment]
@brtnfld What is preventing this branch to be merged ?

**Comment by brtnfld** [comment]
Let me look at agian, I'll do it this week.

**Comment by MicK7** [comment]
🆙

**Review by brtnfld — APPROVED** [review]

**Comment by pini-gh** [comment]
Thanks!

**Pull request merged**